h/services/user_delete.py

```diff
@@ -1,70 +1,315 @@
-            annotations=self._db.query(Annotation).filter_by(userid=user.userid).all()
-        )
-
-        # Delete or remove our link to groups we've created
-        for group, annotations_by_other_users in self._db.execute(
-            # pylint:disable=not-callable
-            sa.select(Group, sa.func.count(Annotation.id))
-            .where(Group.creator == user)
-            .outerjoin(
-                Annotation,
-                sa.and_(
-                    Annotation.groupid == Group.pubid, Annotation.userid != user.userid
-                ),
+        # We can't just purge all the user's data right now because for users
+        # with a lot of data this takes too long for an HTTP request and the
+        # request times out.
+        #
+        # So add a job to purge the user's data in the background.
+        self.db.add(
+            Job(
+                name=self.job_name,
+                priority=0,
+                tag="UserDeleteService.delete_user",
+                kwargs={"userid": user.userid},
             )
-            .group_by(Group.id)
-        ):
-            if annotations_by_other_users:
-                group.creator = None
-            else:
-                self._db.delete(group)
+        )
 
-        self._db.add(
+        # Record the deletion for record-keeping purposes.
+        self.db.add(
             UserDeletion(
                 userid=user.userid,
                 requested_by=requested_by.userid,
                 tag=tag,
                 registered_date=user.registered_date,
-                num_annotations=self._db.scalar(
-                    sa.select(
-                        sa.func.count(Annotation.id)  # pylint:disable=not-callable
+                num_annotations=self.db.scalar(
+                    select(
+                        func.count(Annotation.id)  # pylint:disable=not-callable
                     ).where(Annotation.userid == user.userid)
                 ),
             )
         )
 
-        self._db.delete(user)
+    def purge_deleted_users(self, limit=1000):
+        """Incrementally purge data of users who've been marked as deleted."""
+        jobs = self.job_queue.get(self.job_name, limit)
+
+        if not jobs:
+            log.info("There are no '%s' jobs", self.job_name)
+            return
+
+        completed_jobs = []
+        purger = UserPurger(self.job_queue, LimitedWorker(self.db, limit))
+
+        for job in jobs:
+            userid = job.kwargs.get("userid")
+
+            if not userid:
+                log.info("Invalid '%s' job: %s", self.job_name, job)
+                completed_jobs.append(job)
+                continue
+
+            user = self.user_svc.fetch(userid)
+
+            if user is None:
+                log.info("Couldn't fetch user: %s", userid)
+                completed_jobs.append(job)
+                continue
+
+            try:
+                purger.delete_authtickets(user)
```

Comment:
So this task will go as far down this list as the limit permits.

When finishing via the break the transactions will be committed and next time around it will continue down the list.